File: otp_protection/lib/core/services/message_parser.dart
import 'package:otp_protection/core/models/sms_message.dart';
import 'package:otp_protection/core/utils/otp_result.dart';

class MessageParser {
  static const _otpKeywords = [
    'otp',
    'verification code',
    'verification',
    'security code',
    'one-time password',
    'passcode',
    'authentication code',
    'login code',
    'code',
  ];

  static final _otpRegex = RegExp(r'\b\d{4,8}\b');

  OtpResult parse(SmsMessage sms) {
    final normalized = sms.content.toLowerCase();

    final hasKeyword = _otpKeywords.any(
      normalized.contains,
    );

    if (!hasKeyword) {
      return OtpResult(
        containsOtp: false,
        otp: null,
      );
    }

    final match = _otpRegex.firstMatch(sms.content);

    if (match == null) {
      return OtpResult(
        containsOtp: false,
        otp: null,
      );
    }

    return OtpResult(
      containsOtp: true,
      otp: match.group(0),
    );
  }
}

File: otp_protection/android/app/src/main/kotlin/com/example/otp_protection/message_parser.kt

package com.example.otp_protection
class SmsParser {

    private val otpKeywords = listOf(
        "otp",
        "verification code",
        "verification",
        "security code",
        "one-time password",
        "passcode",
        "authentication code",
        "login code"
    )

    private val otpRegex =
        Regex("\\b\\d{4,8}\\b")

    fun containsOtp(
        message: String
    ): Boolean {

        val normalized =
            message.lowercase()

        val hasKeyword =
            otpKeywords.any {
                normalized.contains(it)
            }

        if (!hasKeyword) {
            return false
        }

        return otpRegex.containsMatchIn(message)
    }
}
